test/rxd/testutils.py

```diff
@@ -85,7 +84,10 @@ def collect_data(h, rxd, data, save_path, num_record=10):
 def compare_data(data):
     """compares the test data with the correct data"""
 
-    rlen = data["rlen"]
+    if "rlen" in data:
```

Comment:
This turns out to needed for ```pytest -s test/rxd/ecs/test_ecs_pure_diffusion.py``` where a bunch of subtests, e.g. ```pytest -s test/rxd/ecs/test_ecs_pure_diffusion.py::test_ecs_diffusion_2d``` generate no data.
@adamjhn Let me know if these tests are supposed to generate substantive data.

Reply:
@adamjhn I added the following fragment:
```
+++ b/test/rxd/ecs/test_ecs_pure_diffusion.py
@@ -8,6 +8,10 @@ def ecs_diffusion(neuron_instance):
     h, rxd, data, save_path = neuron_instance
 
     def make_model(nx, ny, nz, alpha, lambd, d=1.0, perm=None):
+        if len(list(h.allsec())) == 0:
+            # Only needed to gather data.
+            # Persists until cleaned up by neuron_nosave_instance
+            h("create dummy")
         dx = 10
```
and, now, the data produces  with ```pytest -s test/rxd/ecs/test_ecs_pure_diffusion.py --save test/rxd/testdata/tests```  (copying into test/rxd/testdata/test) gives
```
~/neuron/rxd/test/rxd/testdata/test$ git status .
On branch hines/data-fix
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   ecs_diffusion_1d.dat
	modified:   ecs_diffusion_1d_alpha.dat
	modified:   ecs_diffusion_1d_cvode.dat
	modified:   ecs_diffusion_1d_cvode_alpha.dat
	modified:   ecs_diffusion_1d_cvode_tort.dat
	modified:   ecs_diffusion_1d_tort.dat
	modified:   ecs_diffusion_2d.dat
	modified:   ecs_diffusion_2d_alpha.dat
	modified:   ecs_diffusion_2d_cvode.dat
	modified:   ecs_diffusion_2d_cvode_alpha.dat
	modified:   ecs_diffusion_2d_cvode_tort.dat
	modified:   ecs_diffusion_2d_tort.dat
	modified:   ecs_diffusion_3d_alpha_state.dat
	modified:   ecs_diffusion_3d_inhom_tort.dat
	modified:   ecs_diffusion_3d_inhom_tort_cvode.dat
	modified:   ecs_diffusion_3d_tort_state.dat
	modified:   ecs_diffusion_anisotropic.dat
	modified:   ecs_diffusion_cvode_anisotropic.dat

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	ecs_diffusion_0d.dat
	ecs_diffusion_0d_alpha.dat
	ecs_diffusion_0d_cvode.dat
	ecs_diffusion_0d_cvode_alpha.dat
	ecs_diffusion_0d_cvode_tort.dat
	ecs_diffusion_0d_tort.dat
```
That everything was modified makes sense to me because there are now three extra voltage states because of the dummy section. But note the untracked files are still small:
```
~/neuron/rxd/test/rxd/testdata/test$ l ecs_diffusion_0d*
-rw-rw-r-- 1 hines hines 240 Nov  5 17:37 ecs_diffusion_0d_alpha.dat
-rw-rw-r-- 1 hines hines  48 Nov  5 17:37 ecs_diffusion_0d_cvode_alpha.dat
-rw-rw-r-- 1 hines hines  48 Nov  5 17:37 ecs_diffusion_0d_cvode.dat
-rw-rw-r-- 1 hines hines  48 Nov  5 17:37 ecs_diffusion_0d_cvode_tort.dat
-rw-rw-r-- 1 hines hines 240 Nov  5 17:37 ecs_diffusion_0d.dat
-rw-rw-r-- 1 hines hines 240 Nov  5 17:37 ecs_diffusion_0d_tort.dat
```
I did write json files as well and:
```
~/neuron/rxd/test/rxd/testdata/tests$ cat ecs_diffusion_0d_cvode_tort.dat.json
{
  "record_count": 2,
  "data": [
    0.0,
    1000.0,
    1.0,
    3.4322590648824254,
    1000.0,
    1.0
  ],
  "rlen": 3
}
``` 
One of the larger files is
```
$ l test/ecs_diffusion_3d_tort_state.dat
-rw-rw-r-- 1 hines hines 213040 Nov  5 18:37 test/ecs_diffusion_3d_tort_state.dat
hines@hines-ThinkStation-P5:~/neuron/rxd/test/rxd/testdata$ l tests/ecs_diffusion_3d_tort_state.dat
-rw-rw-r-- 1 hines hines 213120 Nov  5 17:28 tests/ecs_diffusion_3d_tort_state.dat
```
So that is an example that had data anyway.
I must be missing something. I thought that without a section, there would be no data gathered. Did these other tests in fact have sections?

Reply:
You're right. I've replaced the extra_scatter_gather data for test_ecs_pure_diffusion with the point process data. The latter was forced to be generated by creating a dummy section for that test.